### PR TITLE
Process injected mouse events

### LINKED ISSUE #7480: NVDA not reading the objects under the mouse cursor when mouse emulator is used.
Some of our clients use emulator freeware such as “JoyToKey” to allow joystick to mimic a mouse, so they can control the mouse cursor with a joystick. It is found that NVDA would not read aloud the object or text under the mouse cursor when such emulator is used to control the mouse cursor. Do you have any suggestion on how we can make NVDA works with these mouse emulators? 

Thanks so much!

System configuration:

NVDA version:  2017.2 Installed
This issue exists in all versions. 

Windows version:　Windows 7  6.1.7601

Name and version of other software in use when reproducing the issue:
JoyToKey (January 2016 version)
Website:　http://www.oneswitch.org.uk/2/I/JoyToKey/JoyToKey.htm

### LINKED ISSUE #8452: Let NVDA process injected mouse events if desired
### Steps to reproduce:
1. Using remote control software like TeamViewer, control the machine of an NVDA user
2. Move the mouse while the remote system has mouse tracking enabled

### Actual behavior:
NVDA on the remote system doesn't echo what is under the mouse.

### Expected behavior:
NVDA on the remote system echoes what is under the mouse.

### Technical details
NVDA is simply ignoring injected mouse events. I propose a system that is similar to how injection is handled for the keyboard.

#### NVDA Installed/portable/running from source:
Both installed and running from source

#### NVDA version:
Latest next and master snapshots, but honestly, this isn't very relevant.

#### Windows version:
Windows 10 april update

#### Name and version of other software in use when reproducing the issue:
TeamViewer 10, ConnectWise Control Remote Support

### PR DESCRIPTION
**Pull request by LeonarddeR**
### Link to issue number:
Closes #8452 
Fixes #7480

### Summary of the issue:
NVDA silently ignores mouse events that are injected by third party applications, particularly remote control software such as TeamViewer. When mouse tracking is enabled, NVDA does not announce what is under the mouse when a TeamViewer partner is controlling it.

### Description of how this pull request fixes the issue:
1. This pr adds a new mouse option to enable/disable the processing of injected mouse control, similar to the "handle keys from other applications" keyboard option. This option is enabled by default.
2. Since NVDA also executes mouse events by itself, events executed with winUser.mouse_event are still ignored. To accomplish this without having winUser tamper with a boolean variable in mouseHandler, there is now a ignoreINjection context manager in mouseHandler that deals with this. For consistency reasons, i also added an ignoreINjection context manager to keyboardHandler.

### Testing performed:
Tested injected mouse events by having a colleague control my system remotely via TeamViewer. His mouse movement was announced by NVDA.

### Known issues with pull request:
Mouse movement caused by SetCursorPos, which is strictly spoken only moving the mouse cursor, is still ignored by NVDA. We probably have to hook SetCursorPos if we want to support this, but I don't think that this is relevant. IN any case, it falls outside the scope of this pr.

### Change log entry:
* New features
    + Added an option to NVDA's mouse settings to make NVDA handle situations where the mouse is controlled by another application. (#8452)
        - This will allow NVDA to track the mouse when a system is controlled remotely using TeamViewer or other remote control software.

### CODE REVIEW

File: source/keyboardHandler.py [hunk 1]
```diff
@@ -60,6 +61,14 @@
 #: Whether the sticky NVDA modifier is locked.
 stickyNVDAModifierLocked = False
 
+@contextmanager
+def ignoreInjection():
+	"""Context manager that allows ignoring injected keys temporarily by using a with statement."""
+	global ignoreInjected
```
Comment (feerrenrut) [inline source/keyboardHandler.py]: This won't be thread safe. If multiple threads use this context manager at once, the value of ignoreInjected could be squashed.

File: source/gui/settingsDialogs.py [hunk 2]
```diff
@@ -1353,13 +1353,20 @@ def makeSettings(self, settingsSizer):
 		self.audioDetectBrightnessCheckBox=sHelper.addItem(wx.CheckBox(self,label=audioDetectBrightnessText))
 		self.audioDetectBrightnessCheckBox.SetValue(config.conf["mouse"]["audioCoordinates_detectBrightness"])
 
+		# Translators: This is the label for a checkbox in the
+		# mouse settings panel.
+		handleInjectedMouseControlText = _("Handle mouse control from other &applications")
```
Comment (feerrenrut) [inline source/gui/settingsDialogs.py]: I don't think the naming of this setting conveys what it does. Maybe something like "Ignore mouse movement triggered by other applications"

File: source/winUser.py [hunk 3]
```diff
@@ -430,7 +430,11 @@ def keybd_event(*args):
 	return user32.keybd_event(*args)
 
 def mouse_event(*args):
-	return user32.mouse_event(*args)
+	# #8452: NVDA should ignore mouse events that are injected by itself.
+	# Import late to avoid circular import
+	import mouseHandler
+	with mouseHandler.ignoreInjection():
+		return user32.mouse_event(*args)
```
Comment (feerrenrut) [inline source/winUser.py]: I don't think that we can rely on the timing of receiving these events. Currently this design relies on the events being received while this contextManager is active.

**Review by feerrenrut — CHANGES_REQUESTED** [review]
I'm a little worried about the robustness of this approach.

File: source/gui/settingsDialogs.py (hunk 2 above)
Comment (leonardder) [reply source/gui/settingsDialogs.py]: I started with the word movement, but the problem with that word is that strictly spoken, mouse clicks aren't movement either.

File: source/winUser.py (hunk 3 above)
Comment (leonardder) [reply source/winUser.py]: Hmmmm, a similar approach is used in keyboardHandler.KeyboardInputGesture.send, though there, wx.Yield is called before closing the context manager.

Comment (feerrenrut) [reply source/winUser.py]: I have a feeling that code relies on a sleep, which I'm not a fan of either! Really we need a way to identify that a future received event is the one we planned to ignore, and only ignore that one.

File: source/gui/settingsDialogs.py (hunk 2 above)
Comment (leonardder) [reply source/gui/settingsDialogs.py]: Hmm, ignore mouse input would turn the boolean upside down and causes a symmetrical conflict with handle keys from other applications  in the keyboard settings.

How about "Process mouse input from other applications"

File: source/winUser.py (hunk 3 above)
Comment (leonardder) [reply source/winUser.py]: Hmm yes, this could be a cause of #8020.
Note that similar to wx.Yield, we also have wx.YieldIfNeeded.
I"m still not sure what to do here. @michaeldcurran or @jcsteh, thoughts?

**Comment by leonardder** [comment]
@josephsl: Could you elaborate on differences between WX Python 3 and 4 regarding wx.Yield? This code depends on wx.Yield, so it might make sense to change this as part of this pr at least in this area of the code.

**Comment by josephsl** [comment]
Hi, last time I checked, wx.Yield in wxPython 4 calls wx.AppConsole.Yield. I, @JCSTeh, and wxPython developer had a discussion about this a while back. Thanks.

**Comment by leonardder** [comment]
It might make sense to implement #8679 as part of this pr.

File: source/gui/settingsDialogs.py (hunk 2 above)
Comment (leonardder) [reply source/gui/settingsDialogs.py]: After considering this a little more, I think that I agree now with the wording using "ignore" now. May be it makes sense to reword the keyboard settings option as well?

File: source/winUser.py (hunk 3 above)
Comment (leonardder) [reply source/winUser.py]: I discovered that there's also `wx.GetApp().SafeYieldFor` which allows a bitmask of events to be provided, limiting the yield to whatever events you provide.

Regarding SafeYield:

> This function is similar to wx.Yield , except that it disables the user input to all program windows before calling wx.AppConsole.Yield and re-enables it again afterwards.

**Comment by leonardder** [comment]
@feerrenrut: I'm a bit unsure as what to do to make this pr to a success now. I and others would really appreciate this being included in a next release of NVDA>

**Comment by michaelDCurran** [comment]
I can look into this.

File: source/config/configSpec.py [hunk 4]
```diff
@@ -101,6 +101,7 @@
 	audioCoordinates_minPitch = integer(default=220)
 	audioCoordinates_maxPitch = integer(default=880)
 	reportMouseShapeChanges = boolean(default=false)
+	ignoreINjectedMouseInput = boolean(default=false)
```
Comment (michaelDCurran) [inline source/config/configSpec.py]: Should not be a capital N in INjected.

File: source/gui/settingsDialogs.py [hunk 5]
```diff
@@ -1363,13 +1363,20 @@ def makeSettings(self, settingsSizer):
 		self.audioDetectBrightnessCheckBox=sHelper.addItem(wx.CheckBox(self,label=audioDetectBrightnessText))
 		self.audioDetectBrightnessCheckBox.SetValue(config.conf["mouse"]["audioCoordinates_detectBrightness"])
 
+		# Translators: This is the label for a checkbox in the
+		# mouse settings panel.
+		ignoreINjectedMouseInputText = _("Ignore mouse input from other &applications")
```
Comment (michaelDCurran) [inline source/gui/settingsDialogs.py]: Again INjected should not have capital N

File: source/gui/settingsDialogs.py [hunk 6]
```diff
@@ -1363,13 +1363,20 @@ def makeSettings(self, settingsSizer):
 		self.audioDetectBrightnessCheckBox=sHelper.addItem(wx.CheckBox(self,label=audioDetectBrightnessText))
 		self.audioDetectBrightnessCheckBox.SetValue(config.conf["mouse"]["audioCoordinates_detectBrightness"])
 
+		# Translators: This is the label for a checkbox in the
+		# mouse settings panel.
+		ignoreINjectedMouseInputText = _("Ignore mouse input from other &applications")
+		self.ignoreINjectedMouseInputCheckBox=sHelper.addItem(wx.CheckBox(self,label=ignoreINjectedMouseInputText))
+		self.ignoreINjectedMouseInputCheckBox.SetValue(config.conf["mouse"]["ignoreINjectedMouseInput"])
```
Comment (michaelDCurran) [inline source/gui/settingsDialogs.py]: Okay, looks like a bad find and replace ;)

File: source/winUser.py [hunk 7]
```diff
@@ -433,7 +433,11 @@ def keybd_event(*args):
 	return user32.keybd_event(*args)
 
 def mouse_event(*args):
-	return user32.mouse_event(*args)
+	# #8452: NVDA should ignore mouse events that are injected by itself.
+	# Import late to avoid circular import
+	import mouseHandler
+	with mouseHandler.ignoreInjection():
```
Comment (michaelDCurran) [inline source/winUser.py]: I do not like NVDA specific state being affected in winUser. winUser should really just be a module that pythonizes win32 APIs. Rather add a wrapper function in mouseHandler or a new mouseUtils module or something.

**Review by michaelDCurran — CHANGES_REQUESTED** [review]
`KeyboardInputGesture.send` requires a sleep and wx.yield because we want to guarantee that by the time that send returns, the keyboard input generated has been injected and NVDA has received and processed it.  As we do not have scripts for mouse input currently, this is not relevant for the mouse. 
From what I can read on msdn, low-level keyboard hooks will be run within the call to keybd_event, so to low-level mouse hooks will be run within mouse_event. Thus we can guarantee that the related ignore input variable we are setting with the context manager will be correct when the hooks run. 
Perhaps we should look into replacing wx.yield in keyboardHandler, but I feel this is not something for this pr.
In short, once the other review actions in my review here are addressed, I am happy to approve this.

File: source/winUser.py (hunk 7 above)
Comment (leonardder) [reply source/winUser.py]: Valid point there. It required some changes to get this right, though.

**Review by michaelDCurran — APPROVED** [review]

**Comment by michaelDCurran** [comment]
@leonardder Are you happy for this to be merged to master now?

**Comment by leonardder** [comment]
Sure!
